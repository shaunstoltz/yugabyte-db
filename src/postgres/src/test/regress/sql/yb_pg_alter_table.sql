--
-- ALTER_TABLE
--

-- Clean up in case a prior regression run failed
SET client_min_messages TO 'warning';
DROP ROLE IF EXISTS regress_alter_table_user1;
RESET client_min_messages;

CREATE USER regress_alter_table_user1;

--
-- lock levels
--
drop type lockmodes;
create type lockmodes as enum (
 'SIReadLock'
,'AccessShareLock'
,'RowShareLock'
,'RowExclusiveLock'
,'ShareUpdateExclusiveLock'
,'ShareLock'
,'ShareRowExclusiveLock'
,'ExclusiveLock'
,'AccessExclusiveLock'
);

drop view my_locks;
create or replace view my_locks as
select case when c.relname like 'pg_toast%' then 'pg_toast' else c.relname end, max(mode::lockmodes) as max_lockmode
from pg_locks l join pg_class c on l.relation = c.oid
where virtualtransaction = (
        select virtualtransaction
        from pg_locks
        where transactionid = txid_current()::integer)
and locktype = 'relation'
and relnamespace != (select oid from pg_namespace where nspname = 'pg_catalog')
and c.relname != 'my_locks'
group by c.relname;

create table alterlock (f1 int primary key, f2 text);
insert into alterlock values (1, 'foo');
create table alterlock2 (f3 int primary key, f1 int);
insert into alterlock2 values (1, 1);

begin; alter table alterlock alter column f2 set statistics 150;
select * from my_locks order by 1;
rollback;

begin; alter table alterlock cluster on alterlock_pkey;
select * from my_locks order by 1;
commit;

begin; alter table alterlock set without cluster;
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (fillfactor = 100);
select * from my_locks order by 1;
commit;

begin; alter table alterlock reset (fillfactor);
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (toast.autovacuum_enabled = off);
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (autovacuum_enabled = off);
select * from my_locks order by 1;
commit;

begin; alter table alterlock alter column f2 set (n_distinct = 1);
select * from my_locks order by 1;
rollback;

-- test that mixing options with different lock levels works as expected
begin; alter table alterlock set (autovacuum_enabled = off, fillfactor = 80);
select * from my_locks order by 1;
commit;

begin; alter table alterlock alter column f2 set storage extended;
select * from my_locks order by 1;
rollback;

-- TODO(jason): uncomment when doing issue #9106
-- begin; alter table alterlock alter column f2 set default 'x';
-- select * from my_locks order by 1;
-- rollback;

--
-- typed tables: OF / NOT OF
--

CREATE TYPE tt_t0 AS (z inet, x int, y numeric(8,2));
ALTER TYPE tt_t0 DROP ATTRIBUTE z;
CREATE TABLE tt0 (x int NOT NULL, y numeric(8,2));	-- OK
CREATE TABLE tt1 (x int, y bigint);					-- wrong base type
CREATE TABLE tt2 (x int, y numeric(9,2));			-- wrong typmod
CREATE TABLE tt3 (y numeric(8,2), x int);			-- wrong column order
CREATE TABLE tt4 (x int);							-- too few columns
CREATE TABLE tt5 (x int, y numeric(8,2), z int);	-- too few columns
CREATE TABLE tt6 () INHERITS (tt0);					-- can't have a parent
CREATE TABLE tt7 (x int, q text, y numeric(8,2)) WITH OIDS;
ALTER TABLE tt7 DROP q;								-- OK

ALTER TABLE tt0 OF tt_t0;
ALTER TABLE tt1 OF tt_t0;
ALTER TABLE tt2 OF tt_t0;
ALTER TABLE tt3 OF tt_t0;
ALTER TABLE tt4 OF tt_t0;
ALTER TABLE tt5 OF tt_t0;
ALTER TABLE tt6 OF tt_t0;
ALTER TABLE tt7 OF tt_t0;

CREATE TYPE tt_t1 AS (x int, y numeric(8,2));
ALTER TABLE tt7 OF tt_t1;			-- reassign an already-typed table
ALTER TABLE tt7 NOT OF;
\d tt7
