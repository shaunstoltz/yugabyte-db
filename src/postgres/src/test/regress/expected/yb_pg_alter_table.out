--
-- ALTER_TABLE
--
-- Clean up in case a prior regression run failed
SET client_min_messages TO 'warning';
DROP ROLE IF EXISTS regress_alter_table_user1;
RESET client_min_messages;
CREATE USER regress_alter_table_user1;
--
-- lock levels
--
drop type lockmodes;
ERROR:  type "lockmodes" does not exist
create type lockmodes as enum (
 'SIReadLock'
,'AccessShareLock'
,'RowShareLock'
,'RowExclusiveLock'
,'ShareUpdateExclusiveLock'
,'ShareLock'
,'ShareRowExclusiveLock'
,'ExclusiveLock'
,'AccessExclusiveLock'
);
drop view my_locks;
ERROR:  view "my_locks" does not exist
create or replace view my_locks as
select case when c.relname like 'pg_toast%' then 'pg_toast' else c.relname end, max(mode::lockmodes) as max_lockmode
from pg_locks l join pg_class c on l.relation = c.oid
where virtualtransaction = (
        select virtualtransaction
        from pg_locks
        where transactionid = txid_current()::integer)
and locktype = 'relation'
and relnamespace != (select oid from pg_namespace where nspname = 'pg_catalog')
and c.relname != 'my_locks'
group by c.relname;
create table alterlock (f1 int primary key, f2 text);
insert into alterlock values (1, 'foo');
create table alterlock2 (f3 int primary key, f1 int);
insert into alterlock2 values (1, 1);
begin; alter table alterlock alter column f2 set statistics 150;
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set statistics 150;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
begin; alter table alterlock cluster on alterlock_pkey;
ERROR:  ALTER TABLE CLUSTER not supported yet
LINE 1: alter table alterlock cluster on alterlock_pkey;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set without cluster;
ERROR:  ALTER TABLE SET WITHOUT CLUSTER not supported yet
LINE 1: alter table alterlock set without cluster;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (fillfactor = 100);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (fillfactor = 100);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock reset (fillfactor);
ERROR:  ALTER TABLE RESET not supported yet
LINE 1: alter table alterlock reset (fillfactor);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (toast.autovacuum_enabled = off);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (toast.autovacuum_enabled = off);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (autovacuum_enabled = off);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (autovacuum_enabled = off);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock alter column f2 set (n_distinct = 1);
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set (n_distinct = 1);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
-- test that mixing options with different lock levels works as expected
begin; alter table alterlock set (autovacuum_enabled = off, fillfactor = 80);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (autovacuum_enabled = off, fillfac...
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock alter column f2 set storage extended;
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set storage extended;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
-- TODO(jason): uncomment when doing issue #9106
-- begin; alter table alterlock alter column f2 set default 'x';
-- select * from my_locks order by 1;
-- rollback;
--
-- typed tables: OF / NOT OF
--
CREATE TYPE tt_t0 AS (z inet, x int, y numeric(8,2));
ALTER TYPE tt_t0 DROP ATTRIBUTE z;
ERROR:  ALTER TYPE DROP ATTRIBUTE not supported yet
LINE 1: ALTER TYPE tt_t0 DROP ATTRIBUTE z;
                         ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1893. Click '+' on the description to raise its priority
CREATE TABLE tt0 (x int NOT NULL, y numeric(8,2));	-- OK
CREATE TABLE tt1 (x int, y bigint);					-- wrong base type
CREATE TABLE tt2 (x int, y numeric(9,2));			-- wrong typmod
CREATE TABLE tt3 (y numeric(8,2), x int);			-- wrong column order
CREATE TABLE tt4 (x int);							-- too few columns
CREATE TABLE tt5 (x int, y numeric(8,2), z int);	-- too few columns
CREATE TABLE tt6 () INHERITS (tt0);					-- can't have a parent
ERROR:  INHERITS not supported yet
LINE 1: CREATE TABLE tt6 () INHERITS (tt0);
                            ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1129. Click '+' on the description to raise its priority
CREATE TABLE tt7 (x int, q text, y numeric(8,2)) WITH OIDS;
ERROR:  OIDs are not supported for user tables.
ALTER TABLE tt7 DROP q;								-- OK
ERROR:  relation "tt7" does not exist
ALTER TABLE tt0 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt0 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt1 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt1 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt2 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt2 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt3 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt3 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt4 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt4 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt5 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt5 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt6 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt6 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt7 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt7 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
CREATE TYPE tt_t1 AS (x int, y numeric(8,2));
ALTER TABLE tt7 OF tt_t1;			-- reassign an already-typed table
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt7 OF tt_t1;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt7 NOT OF;
ERROR:  ALTER TABLE NOT OF not supported yet
LINE 1: ALTER TABLE tt7 NOT OF;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
\d tt7
